xtask/src/main.rs

```diff
@@ -0,0 +1,607 @@
+                ],
+                Some("firmware/panic-probe"),
+                &[],
+            )
+        },
+        "cross",
+        errors,
+    );
+
+    do_test(
+        || {
+            run_command::<&str>(
+                &[
+                    "cargo",
+                    "check",
+                    "--target",
+                    "thumbv6m-none-eabi",
+                    "--features",
+                    "print-rtt",
+                ],
+                Some("firmware/panic-probe"),
+                &[],
+            )
+        },
+        "cross",
+        errors,
+    )
+}
+
+fn test_snapshot(errors: &mut Vec<String>) {
+    println!("🧪 qemu/snapshot");
+    let mut tests = vec![
+        "log",
+        "timestamp",
+        "panic",
+        "assert",
+        "assert-eq",
+        "assert-ne",
+        "unwrap",
+        "defmt-test",
+        "hints",
+    ];
+
+    match rustc_is_nightly() {
+        Ok(is_nightly) => {
+            if is_nightly {
+                tests.push("alloc");
+            }
+        }
+        Err(e) => {
+            eprintln!(
+                "could not determine whether rust compiler is nightly - assuming it's not ({})",
+                e
+            )
+        }
+    }
+
+    let mut features_map = HashMap::new();
+    features_map.insert("alloc", "alloc");
+    let no_features = "";
+
+    for test in &tests {
+        let features = features_map.get(test).unwrap_or(&no_features);
+
+        do_test(
+            || test_single_snapshot(test, features, false),
+            "qemu/snapshot",
+            errors,
+        );
+        do_test(
+            || test_single_snapshot(test, features, true),
+            "qemu/snapshot",
+            errors,
+        );
+    }
+}
+
+fn main() -> Result<(), Vec<String>> {
+    let opt: Options = Options::from_args();
+    let mut all_errors: Vec<String> = vec![];
+
+    // TODO: one could argue that not all test scenarios require installation of targets
+    let added_targets = match install_targets() {
+        Ok(targets) => targets,
+        Err(e) => {
+            panic!("Error while installing required targets: {}", e)
+        }
+    };
+
+    match opt.cmd {
+        TestCommand::TestAll => {
+            test_host(&mut all_errors);
+            test_cross(&mut all_errors);
+            test_snapshot(&mut all_errors);
+            test_book(&mut all_errors);
+            test_lint(&mut all_errors);
+        }
+        TestCommand::TestHost => {
+            test_host(&mut all_errors);
+        }
+        TestCommand::TestCross => {
+            test_cross(&mut all_errors);
+        }
+        TestCommand::TestSnapshot => {
+            test_snapshot(&mut all_errors);
+        }
+        TestCommand::TestBook => {
+            test_book(&mut all_errors);
+        }
+        TestCommand::TestLint => {
+            test_lint(&mut all_errors);
+        }
+    }
+
+    if !opt.keep_targets {
+        uninstall_targets(added_targets);
```

Comment:
Ah, I see, this is why you query the targets. Seems fine then, but I don't think this functionality is necessary.

Reply:
I can still `" "`join them for installing though :D Rust is *all about efficiency* after all

Reply:
(Optional uninstall was suggested by Jorge and from my POV it was a quick+easy win, so I added it)